🚀 1단계 - 예약 기능 구현(1-3단계 - 시간 관리 기능 추가)

junseong94 opened the pull request:
1-3 단계 미션 구현 완료하여 PR 드립니다.
감사합니다.!

junseong94 committed:
실습 페이지 설정 완료

junseong94 committed:
예약 관리 API - 예약조회 API 구현 완료

junseong94 committed:
예약 관리 API - 예약추가 API 구현 완료

junseong94 committed:
예약 관리 API - 예약취소 API 구현 완료

junseong94 committed:
h2 database 설정 완료

junseong94 committed:
refactor : Reservation, ReservationController
create : ReservationResponseDto
update : ReservationDto to ReservationRequestDto

junseong94 committed:
update : 예악저장소 명세 변경(HashMap to Map)

junseong94 committed:
ReservationRepository : 구현 및 테스트 코드 작성

junseong94 committed:
ReservationService 추가 구현
ReservationController 리팩토링
GlobalExceptionHandler 추가
ReservationController 리팩토링 : HashMap 기반 테스트 코드 일부 수정

junseong94 committed:
GlobalExceptionHandler : 부가적인 예외처리 코드 추가
Reservation : id 기준 동등성 부여를 위해 equals, hashCode 추가
ReservationRepository : row Mapper 필드로 추가
ReservationService : existsById를 조건식으로 delete 로직 수정

junseong94 committed:
ReservationTime 도메인 구현

junseong94 committed:
패키지명 수정

junseong94 committed:
init

junseong94 committed:
rebase

junseong94 committed:
테스트 코드 에러 수정

Hyeon9mak commented on src/main/java/roomescape/reservationTime/ReservationTime.java:
```diff
@@ -0,0 +1,51 @@
+    public ReservationTime(Long id, String startAt) {
+        this.id = id;
+        this.startAt = startAt;
+    }
```
getter 와 생성자들의 순서를 정리해주면 가독성이 더욱 올라갈 것 같아요!

Hyeon9mak commented on src/main/java/roomescape/reservationTime/ReservationTimeRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+    public ReservationTimeRequestDto(String startAt) {
+        this.startAt = startAt;
+    }
+
```
마찬가지로 멤버변수, 생성자, getter 순서를 정리해보는 건 어떨까요?

Hyeon9mak commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,27 +1,27 @@
-        this.time = time;
+        this.reservationTime = reservationTime;
     }
 
```
생성자 체이닝을 활용하는 것도 좋은 방법일 것 같아요 😄 

```java
 public Reservation(String name, String date, ReservationTime reservationTime) {
      this(null, name, date, reservationTime);
  }
```

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -21,21 +23,37 @@ public ReservationRepository(JdbcTemplate jdbcTemplate) {
+                "on r.time_id = t.id";
+
         return jdbcTemplate.query(sql, reservationRowMapper);
 
```
java 17 버전 이상을 사용중인 프로젝트네요~

```java
        final String sql = """
                    SELECT r.id as reservation_id,
                    r.name as reservation_name, 
                    r.date as reservation_date, 
                    t.id as time_id, 
                    t.start_at as start_at 
                    FROM reservation as r
                    inner join reservation_time as t
                    on r.time_id = t.id;
                """;

```

Hyeon9mak commented on src/test/java/roomescape/reservation/ReservationControllerTest.java:
```diff
@@ -20,33 +23,64 @@
-        jdbcTemplate.execute("CREATE TABLE reservation(" +
-                "id BIGINT AUTO_INCREMENT, " +
-                "name VARCHAR(255) NOT NULL, " +
-                "date VARCHAR(255), " +
-                "time VARCHAR(255), " +
+        jdbcTemplate.execute("DROP TABLE IF EXISTS reservation");
+        jdbcTemplate.execute("DROP TABLE IF EXISTS reservation_time");
+
+        jdbcTemplate.execute("CREATE TABLE reservation_time (" +
+                "id BIGINT NOT NULL AUTO_INCREMENT, " +
+                "start_at VARCHAR(255) NOT NULL, " +
                 "PRIMARY KEY (id))");
+
+        jdbcTemplate.execute("CREATE TABLE reservation (" +
+                "id BIGINT NOT NULL AUTO_INCREMENT, " +
+                "name VARCHAR(255) NOT NULL, " +
+                "date VARCHAR(255) NOT NULL, " +
+                "time_id BIGINT, " +   // 컬럼 수정
+                "PRIMARY KEY (id))");  // 외래키 제거
```
테스트코드 쪽에만 별도로 FK 를 제거하신 이유가 있을까요?

junseong94 replied:
잠시 제가 미쳤었나봐요... 제가 왜 저렇게 작성했는지 모르겠네요 :( 
FK 추가되는게 맞는데 해당 부분은 무시해주시면 감사하겠습니다 ㅎㅎ;;

Hyeon9mak reviewed (changes requested):
안녕하세요 준성님! 예약기능 구현도 잘 해주셨습니다.
몇 가지 코멘트 남겼으니 확인해주세요!

junseong94 committed:
리팩토링
1. Entity 및 dto 클래스 : getter와 생성자 순서로 정리
2. Reservation Entity : 생성자 체이닝 적용
3. sql 문자열 : 텍스트 블록 적용

Hyeon9mak commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -21,21 +23,40 @@ public ReservationRepository(JdbcTemplate jdbcTemplate) {
 
     public List<Reservation> findAll() {
-        final String sql = "select id, name, date, time from reservation";
+        final String sql = """
+                SELECT r.id as reservation_id, 
+                       r.name as reservation_name, 
```
😉 👍

Hyeon9mak reviewed (approved):
안녕하세요 준성님! 피드백 잘 반영해주셨습니다!
다음 단계로 넘어가시죠 😄
